spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQuery.java

```diff
@@ -108,38 +103,58 @@ else if (this.tree.isDistinct()) {
 	public Object execute(Object[] parameters) {
 		Supplier<StructuredQuery.Builder<?>> queryBuilderSupplier = StructuredQuery::newKeyQueryBuilder;
 		Function<T, ?> mapper = Function.identity();
-		Collector<?, ?, ?> collector;
+		Class returnedType = this.queryMethod.getReturnedObjectType();
+
+		boolean returnedTypeIsNumber = Number.class.isAssignableFrom(returnedType)
+				|| returnedType == int.class || returnedType == long.class;
 
-		if (this.tree.isCountProjection()) {
+		boolean isCountingQuery = this.tree.isCountProjection()
+				|| (this.tree.isDelete() && returnedTypeIsNumber);
+
+		Collector<?, ?, ?> collector = Collectors.toList();
+		if (isCountingQuery) {
 			collector = Collectors.reducing(0, e -> 1, Integer::sum);
 		}
 		else if (this.tree.isExistsProjection()) {
-			collector = Collectors.collectingAndThen(Collectors.counting(), count -> count > 0);
+			collector = Collectors.collectingAndThen(Collectors.counting(),
+					count -> count > 0);
 		}
-		else {
+		else if (!returnedTypeIsNumber) {
 			queryBuilderSupplier = StructuredQuery::newEntityQueryBuilder;
 			mapper = this::processRawObjectForProjection;
-			collector = Collectors.toList();
 		}
 
 		StructuredQuery.Builder<?> structredQueryBuilder = queryBuilderSupplier.get();
 		structredQueryBuilder.setKind(this.datastorePersistentEntity.kindName());
 		applyQueryBody(parameters, structredQueryBuilder);
-		Iterable results = this.datastoreTemplate
+		Iterable rawResults = this.datastoreTemplate
 				.queryKeysOrEntities(structredQueryBuilder.build(),
 				this.entityType);
 
-		Object returned = results == null ? null
-				: StreamSupport.stream(results.spliterator(), false).map(mapper)
+		Object result = StreamSupport.stream(rawResults.spliterator(), false).map(mapper)
 						.collect(collector);
 
-		return this.tree.isExistsProjection() || this.tree.isCountProjection()
-				|| results == null
-						? returned
-						: this.datastoreTemplate.getDatastoreEntityConverter()
-								.getConversions().convertOnRead(returned,
-										this.queryMethod.getCollectionReturnType(),
-										this.queryMethod.getReturnedObjectType());
+		deleteFoundEntities(returnedTypeIsNumber, rawResults);
```

Comment:
It's really odd to see this method call here because it makes it seem like you unconditionally always delete found entities. I know that's not the case, but maybe you should move the `this.tree.isDelete` here.

Reply:
@meltsufin change made, PTAL